### DISCUSSION TITLE
Bagaimana cara improve accuracy pada tuning model CNN ?

### DISCUSSION THREAD

**Discussion by DeanAlexander27** [teamalgoritma/community#634] [Q&A] [answered]
Dear ka Instruktur

apakah ada referensi dalam keras untuk meningkatkan tuning model pada CNN ?

terima kasih

**Comment by idoali** [comment, marked as answer]
Dear Pak Dean,

Terima kasih pak atas pertanyaannya. Sederhananya, bapak bisa melakukan hal-hal dibawah : 
- Menambahkan layer baru pada model. Layer yang bisa ditambahkan bisa berupa CNN atau dense layer. 
- Mengurangi learning rate
- Menambahkan data training
- Mencoba untuk mengganti optimizer
- Membuatnya dari _pretrained-model_ (model deep learning yang sudah setengah jadi. Biasanya kita hanya tinggal melakukan training sedikit lagi di layer terakhirnya).

Berikut sumber yang bisa menjadi referensi penggunaan pretrained model pak : 
https://towardsdatascience.com/step-by-step-guide-to-using-pretrained-models-in-keras-c9097b647b29

Referensi diatas sendiri menggunakan python. Untuk fungsi yang bapak bisa gunakan di R bisa dilihat disini. Tentunya bapak tetap harus melakukan sedikit improvisasi pada kodenya. 
https://tensorflow.rstudio.com/reference/keras/#applications 

Berikut sumber yang bisa bapak baca pak : 
https://stackoverflow.com/questions/70554413/how-to-improve-the-performance-of-cnn-model-for-a-specific-dataset-getting-low

Untuk cara yang sedikit lebih _advanced_, kami juga menemukan artikel ini pak : 
https://towardsdatascience.com/increase-the-accuracy-of-your-cnn-by-following-these-5-tips-i-learned-from-the-kaggle-community-27227ad39554

Semoga bisa menjawab pertanyaannya pak

Regards,
Team Algoritma

**Reply by DeanAlexander27** [reply]
Dear ka @idoali 
saya cukup kebingungan ketika menggunakan vgg16, ketika saya sudah membuat model sebagai berikut : 

tensorflow::tf$random$set_seed(123)

model_big <- keras_model_sequential() %>% 
  
  # First convolutional layer
  layer_conv_2d(filters = 16,
                kernel_size = c(3,3), # 3 x 3 filters
                padding = "same",
                activation = "relu",
                input_shape = c(target_size, 3)
                ) %>%
  
  # First convolutional layer
  layer_conv_2d(filters = 16,
                kernel_size = c(3,3), # 3 x 3 filters
                padding = "same",
                activation = "relu") %>%
  
  # Max pooling layer
  layer_max_pooling_2d(pool_size = c(2,2)) %>% 
  
  # Second convolutional layer
  layer_conv_2d(filters = 32,
                kernel_size = c(3,3), # 3 x 3 filters
                padding = "same",
                activation = "relu"
                ) %>%
  
  # Second convolutional layer
  layer_conv_2d(filters = 32,
                kernel_size = c(3,3), # 3 x 3 filters
                padding = "same",
                activation = "relu"
                ) %>%
  
  # Max pooling layer
  layer_max_pooling_2d(pool_size = c(2,2)) %>% 

  # Third convolutional layer
  layer_conv_2d(filters = 64,
                kernel_size = c(3,3),
                padding = "same",
                activation = "relu"
                ) %>%
  
  # Third convolutional layer
  layer_conv_2d(filters = 64,
                kernel_size = c(3,3),
                padding = "same",
                activation = "relu"
                ) %>%
  
  # Third convolutional layer
  layer_conv_2d(filters = 64,
                kernel_size = c(3,3),
                padding = "same",
                activation = "relu"
                ) %>%

  # Max pooling layer
  layer_max_pooling_2d(pool_size = c(2,2)) %>% 
  
   # Fourth convolutional layer
  layer_conv_2d(filters = 128,
                kernel_size = c(3,3),
                padding = "same",
                activation = "relu"
                ) %>%
  
   # Fourth convolutional layer
  layer_conv_2d(filters = 128,
                kernel_size = c(3,3),
                padding = "same",
                activation = "relu"
                ) %>%
  
   # Fourth convolutional layer
  layer_conv_2d(filters = 128,
                kernel_size = c(3,3),
                padding = "same",
                activation = "relu"
                ) %>%
  
  # Max pooling layer
  layer_max_pooling_2d(pool_size = c(2,2)) %>%
  
  # Fifth convolutional layer
  layer_conv_2d(filters = 256,
                kernel_size = c(3,3),
                padding = "same",
                activation = "relu"
                ) %>%
  
  # Fifth convolutional layer
  layer_conv_2d(filters = 256,
                kernel_size = c(3,3),
                padding = "same",
                activation = "relu"
                ) %>%
  
  # Fifth convolutional layer
  layer_conv_2d(filters = 256,
                kernel_size = c(3,3),
                padding = "same",
                activation = "relu"
                ) %>%
  
  # Max pooling layer
  layer_max_pooling_2d(pool_size = c(2,2)) %>% 

  # Flattening layer
  layer_flatten() %>% 
  
  # Dense layer
  layer_dense(units = 256,
              activation = "relu") %>%
  
  # Dense layer
  layer_dense(units = 128,
              activation = "relu") %>%
  
  # Dense layer
  layer_dense(units = 64,
              activation = "relu") %>%
  
  # Output layer
  layer_dense(name = "Output",
              units = 3, 
              activation = "softmax") 

model_big

model_big %>% 
  compile(
    loss = "categorical_crossentropy",
    optimizer = optimizer_adam(lr = 0.001),
    metrics = "accuracy"
  )

# Fit data into model
history1 <- model_big %>% 
  fit_generator(
  # training data
  train_image_array_gen,

  # training epochs
  steps_per_epoch = as.integer(train_samples / batch_size), 
  epochs = 30, 
  
  # validation data
  validation_data = val_image_array_gen,
  validation_steps = as.integer(valid_samples / batch_size)
)

plot(history1)

dimana ini masuk ke model_big, dan untuk vgg16 , saya melakukan application.vgg16(include_top=F,input_shape = c(target_size,3), weight="imaginet)
pertanyaan saya. nanti berarti saya update architecture kembali dengan model_big %>% keras_model_sequential() %>%vgg kah ?
